File: EasyMatMarker/Go/EasyMatMark.go
package main

import (
	"fmt"
	"time"
	"math/rand/v2"
)

const ROW = 128
const COL = 128
const SIZE = ROW * COL

func add(mat []float64, val float64) {
	for i := range mat {
		mat[i] += val
	}
}

func div(mat []float64, val float64) {
	for i := range mat {
		mat[i] /= val
	}
}

func mul(a []float64, b []float64, c []float64) {
	for row := 0; row < ROW; row++ {
		for col := 0; col < COL; col++ {
			v := 0.
			for i := 0; i < COL; i++ {
				v += a[row*COL+i] * b[i*COL+col]
			}
			c[row*COL+col] = v
		}
	}
}

func main() {
	a := make([]float64, SIZE)
	b := make([]float64, SIZE)
	c := make([]float64, SIZE)
	sumValue := 0.

	for i := range a {
		// a[i] = float64(i+1.) / SIZE
		a[i] = rand.Float64()
	}
	a[128 * 128 - 1] = 1.0

	for i := range b {
		b[i] = 1.
	}

	tp1 := time.Now()

	for i := 0; i < 1000; i++ {
		add(b, float64(i))
		mul(a, b, c)

		add(b, float64(i)+1.)
		mul(c, b, a)

		div(a, c[127*COL+127])
	}

	for i := range a {
		sumValue += a[i]
	}

	tp2 := time.Now()

	fmt.Println(float64(tp2.Sub(tp1).Nanoseconds())/1000./1000., sumValue)
}

File: EasyMatMarker/Python/EasyMatMark.py
import time
import random

# a = [float(i + 1) / float(128 * 128) for i in range(128 * 128)]
a = [random.random() for i in range(128 * 128)]
a[127 * 128 + 127] = 1.0
b = [1 for i in range(128 * 128)]
c = [0 for i in range(128 * 128)]
sum_value = 0


def add(result, x):
    for i in range(128 * 128):
        result[i] += x


def div(result, x):
    for i in range(128 * 128):
        result[i] /= x


def mul(a_, b_, result):
    for row in range(128):
        for col in range(128):
            v = 0
            for i in range(128):
                v += a_[row * 128 + i] * b_[i * 128 + col]
            result[row * 128 + col] = v


tp1 = time.time()

for i in range(1000):
    add(b, i)
    mul(a, b, c)

    add(b, i + 1)
    mul(c, b, a)

    div(a, c[127 * 128 + 127])

sum_value = sum(a)

tp2 = time.time()

print((tp2 - tp1) * 1000., sum_value)
